### PR TITLE
endpoint: add rendezvous hashing support

### PR DESCRIPTION
**Pull request by tsenart**
This commit adds rendezvous hashing support to the endpoint package with
the aim to make it the default in a follow up PR. It's currently
protected by an environment variable which I plan to set on dogfood for
testing.

Additionally, it paves the way for increasing the replication factor by
adding a `GetN` method to map, which I used to replace the previous
`exclude` paramater to `Get` (used only by searcher). Actually raising
replication factor for Zoekt will come later.

[Here](
https://gist.github.com/tsenart/bff23f32d7195d162b712befac6c1f27) is the output of the tests, which shows much better distribution,
and illustrates the data movement needed when changing hash function and
cluster sizes.

I think it's OK for us to shepherd this to prod next week and treat it like
and index version bump, and to write some release notes for customers to expect
some downtime in the next upgrade.

### CODE REVIEW

File: internal/endpoint/endpoint.go [hunk 1]
```diff
@@ -196,10 +214,14 @@ func (m *Map) sync(ch chan endpoints, ready chan struct{}) {
 	}
 }
 
-func newConsistentHashMap(keys []string) *hashMap {
+func newConsistentHash(nodes []string) consistentHash {
+	if os.Getenv("SRC_ENDPOINTS_CONSISTENT_HASH") == "rendezvous" {
```
Comment (tsenart) [inline internal/endpoint/endpoint.go:218]: Here's the env var.

**Comment by tsenart** [comment]
The actual rendesvouz hashing implementation is a fork from dgrisky's with a LookupN implementation: https://github.com/sourcegraph/go-rendezvous

**Review by keegancsmith — APPROVED** [review]
:D Did you mean to commit `internal/endpoint/keys.csv.gz`?

**Comment by rmmh** [comment]
I don't think we should check in keys.csv.gz-- it should be possible to synthesize random repo names that have a close enough distribution to actual to be usable for this.

How much better is xxhash for the distribution than simply `hash/fnv`? I doubt it matters, and one fewer dependency is nice.

A way to smoothly slew from one consistent hashing config to another should make changing them or growing the cluster safer.

**Comment by tsenart** [comment]
@rmmh

> I don't think we should check in keys.csv.gz-- it should be possible to synthesize random repo names that have a close enough distribution to actual to be usable for this.

Producing a syntetic list that matches the hash distribution curve of the real list is more effort than I think it's worth it now, so I'll just `t.Skip` this test if we can't find the file on disk.

> How much better is xxhash for the distribution than simply hash/fnv? I doubt it matters, and one fewer dependency is nice.

We had xxhash as a dependency already, but were using a previous version of the module. Regardless, it seems fnv has even better stddev, so I'm choosing that over xxhash, although we use https://github.com/segmentio/fasthash for it, which we also already depended on.

```
=== RUN   TestConsistentHashing
=== RUN   TestConsistentHashing/Distribution
    consistenthash_test.go:114:
          node  0  consistent(crc32ieee)  =>  50580    5.06%  rendezvous(xxhash64)  =>  41574    4.16%  rendezvous(fnv)  =>  41382    4.14%  rendezvous(fnva)  =>  41913  4.19%
          node  1  consistent(crc32ieee)  =>  51467    5.15%  rendezvous(xxhash64)  =>  41787    4.18%  rendezvous(fnv)  =>  41612    4.16%  rendezvous(fnva)  =>  42095  4.21%
          node  2  consistent(crc32ieee)  =>  52529    5.25%  rendezvous(xxhash64)  =>  41332    4.13%  rendezvous(fnv)  =>  41777    4.18%  rendezvous(fnva)  =>  41594  4.16%
          node  3  consistent(crc32ieee)  =>  47609    4.76%  rendezvous(xxhash64)  =>  41868    4.19%  rendezvous(fnv)  =>  41805    4.18%  rendezvous(fnva)  =>  41571  4.16%
          node  4  consistent(crc32ieee)  =>  46530    4.65%  rendezvous(xxhash64)  =>  41708    4.17%  rendezvous(fnv)  =>  42114    4.21%  rendezvous(fnva)  =>  41431  4.14%
          node  5  consistent(crc32ieee)  =>  41092    4.11%  rendezvous(xxhash64)  =>  41925    4.19%  rendezvous(fnv)  =>  41810    4.18%  rendezvous(fnva)  =>  41453  4.15%
          node  6  consistent(crc32ieee)  =>  47135    4.71%  rendezvous(xxhash64)  =>  41442    4.14%  rendezvous(fnv)  =>  41838    4.18%  rendezvous(fnva)  =>  41391  4.14%
          node  7  consistent(crc32ieee)  =>  38015    3.80%  rendezvous(xxhash64)  =>  41649    4.16%  rendezvous(fnv)  =>  41546    4.15%  rendezvous(fnva)  =>  41482  4.15%
          node  8  consistent(crc32ieee)  =>  38814    3.88%  rendezvous(xxhash64)  =>  41441    4.14%  rendezvous(fnv)  =>  41628    4.16%  rendezvous(fnva)  =>  41673  4.17%
          node  9  consistent(crc32ieee)  =>  37810    3.78%  rendezvous(xxhash64)  =>  41806    4.18%  rendezvous(fnv)  =>  41708    4.17%  rendezvous(fnva)  =>  41718  4.17%
          node 10  consistent(crc32ieee)  =>  42952    4.30%  rendezvous(xxhash64)  =>  41801    4.18%  rendezvous(fnv)  =>  41765    4.18%  rendezvous(fnva)  =>  41916  4.19%
          node 11  consistent(crc32ieee)  =>  39827    3.98%  rendezvous(xxhash64)  =>  41403    4.14%  rendezvous(fnv)  =>  41605    4.16%  rendezvous(fnva)  =>  41626  4.16%
          node 12  consistent(crc32ieee)  =>  36546    3.65%  rendezvous(xxhash64)  =>  41481    4.15%  rendezvous(fnv)  =>  41717    4.17%  rendezvous(fnva)  =>  41659  4.17%
          node 13  consistent(crc32ieee)  =>  30845    3.08%  rendezvous(xxhash64)  =>  41669    4.17%  rendezvous(fnv)  =>  41545    4.15%  rendezvous(fnva)  =>  42002  4.20%
          node 14  consistent(crc32ieee)  =>  36085    3.61%  rendezvous(xxhash64)  =>  41452    4.15%  rendezvous(fnv)  =>  41688    4.17%  rendezvous(fnva)  =>  41768  4.18%
          node 15  consistent(crc32ieee)  =>  37388    3.74%  rendezvous(xxhash64)  =>  41758    4.18%  rendezvous(fnv)  =>  41670    4.17%  rendezvous(fnva)  =>  41474  4.15%
          node 16  consistent(crc32ieee)  =>  29243    2.92%  rendezvous(xxhash64)  =>  41873    4.19%  rendezvous(fnv)  =>  41680    4.17%  rendezvous(fnva)  =>  41627  4.16%
          node 17  consistent(crc32ieee)  =>  37588    3.76%  rendezvous(xxhash64)  =>  41532    4.15%  rendezvous(fnv)  =>  41402    4.14%  rendezvous(fnva)  =>  41727  4.17%
          node 18  consistent(crc32ieee)  =>  46946    4.69%  rendezvous(xxhash64)  =>  41518    4.15%  rendezvous(fnv)  =>  41642    4.16%  rendezvous(fnva)  =>  41979  4.20%
          node 19  consistent(crc32ieee)  =>  42027    4.20%  rendezvous(xxhash64)  =>  41947    4.19%  rendezvous(fnv)  =>  41588    4.16%  rendezvous(fnva)  =>  41803  4.18%
          node 20  consistent(crc32ieee)  =>  41840    4.18%  rendezvous(xxhash64)  =>  41683    4.17%  rendezvous(fnv)  =>  41478    4.15%  rendezvous(fnva)  =>  41514  4.15%
          node 21  consistent(crc32ieee)  =>  47259    4.73%  rendezvous(xxhash64)  =>  41603    4.16%  rendezvous(fnv)  =>  41849    4.18%  rendezvous(fnva)  =>  41542  4.15%
          node 22  consistent(crc32ieee)  =>  43725    4.37%  rendezvous(xxhash64)  =>  42145    4.21%  rendezvous(fnv)  =>  41479    4.15%  rendezvous(fnva)  =>  41642  4.16%
          node 23  consistent(crc32ieee)  =>  36149    3.61%  rendezvous(xxhash64)  =>  41604    4.16%  rendezvous(fnv)  =>  41673    4.17%  rendezvous(fnva)  =>  41401  4.14%

        Stats:
          consistent(crc32ieee)  min: 29243  max: 52529  mean: 41666.708  median: 41466  stdev: 6061.576
           rendezvous(xxhash64)  min: 41332  max: 42145  mean: 41666.708  median: 41659  stdev: 197.915
                rendezvous(fnv)  min: 41382  max: 42114  mean: 41666.708  median: 41671  stdev: 157.520
               rendezvous(fnva)  min: 41391  max: 42095  mean: 41666.708  median: 41634  stdev: 196.601
```

**Comment by tsenart** [comment]
> A way to smoothly slew from one consistent hashing config to another should make changing them or growing the cluster safer.

I'm going to first test this as is in dogfood. Since we already have code that keeps a repo in a shard before it's successfully indexed in the new shard we may be fine. If the experiment in dogfood shows the instance is unusable for a long time, I'll look into a function that takes time into consideration (say 1 week) and gradually moves over repos from the old hash function to the new one.

**Pull request merged**